protocol/streaming/grpc/grpc_streaming_manager.go

```diff
@@ -9,30 +13,125 @@ var _ types.GrpcStreamingManager = (*GrpcStreamingManagerImpl)(nil)
 
 // GrpcStreamingManagerImpl is an implementation for managing gRPC streaming subscriptions.
 type GrpcStreamingManagerImpl struct {
+	sync.Mutex
+
+	// orderbookSubscriptions maps subscription IDs to their respective orderbook subscriptions.
+	orderbookSubscriptions map[uint32]*OrderbookSubscription
+	nextId                 uint32
+}
+
+// OrderbookSubscription represents a active subscription to the orderbook updates stream.
+type OrderbookSubscription struct {
+	clobPairIds []uint32
+	srv         clobtypes.Query_StreamOrderbookUpdatesServer
+	finished    chan bool
 }
 
 func NewGrpcStreamingManager() *GrpcStreamingManagerImpl {
-	return &GrpcStreamingManagerImpl{}
+	return &GrpcStreamingManagerImpl{
+		orderbookSubscriptions: make(map[uint32]*OrderbookSubscription),
+	}
 }
 
 func (sm *GrpcStreamingManagerImpl) Enabled() bool {
 	return true
 }
 
 // Subscribe subscribes to the orderbook updates stream.
+// This function returns a channel that is used to signal termination when an error occurs.
 func (sm *GrpcStreamingManagerImpl) Subscribe(
 	req clobtypes.StreamOrderbookUpdatesRequest,
 	srv clobtypes.Query_StreamOrderbookUpdatesServer,
 ) (
 	finished chan bool,
 	err error,
 ) {
-	return nil, nil
+	clobPairIds := req.GetClobPairId()
+
+	// Perform some basic validation on the request.
+	if len(clobPairIds) == 0 {
+		return nil, clobtypes.ErrInvalidGrpcStreamingRequest
+	}
+
+	finished = make(chan bool)
+	subscription := &OrderbookSubscription{
+		clobPairIds: clobPairIds,
+		srv:         srv,
+		finished:    finished,
+	}
+
+	sm.Lock()
+	defer sm.Unlock()
+
+	sm.orderbookSubscriptions[sm.nextId] = subscription
+	sm.nextId++
+
+	return finished, nil
 }
 
 // SendOrderbookUpdates groups updates by their clob pair ids and
 // sends messages to the subscribers.
 func (sm *GrpcStreamingManagerImpl) SendOrderbookUpdates(
-	updates *clobtypes.OffchainUpdates,
+	offchainUpdates *clobtypes.OffchainUpdates,
 ) {
+	// Group updates by clob pair id.
+	updates := make(map[uint32]*clobtypes.OffchainUpdates)
+	for _, message := range offchainUpdates.Messages {
+		clobPairId := message.OrderId.ClobPairId
+		if _, ok := updates[clobPairId]; !ok {
+			updates[clobPairId] = clobtypes.NewOffchainUpdates()
+		}
+		updates[clobPairId].Messages = append(updates[clobPairId].Messages, message)
+	}
+
+	// Unmarshal messages to v1 updates.
+	v1updates := make(map[uint32][]ocutypes.OffChainUpdateV1)
+	for clobPairId, update := range updates {
+		v1update, err := GetOffchainUpdatesV1(update)
+		if err != nil {
+			panic(err)
```

Comment:
i believe golang kills the whole process on a panic. Would this lead to the whole validator going bye bye? Do we want to have a recover?

Reply:
yeah I was trying to see what makes sense here - 

based on my understanding, MMs will run their own full nodes and these full nodes are only used for streaming information to their trading strategies. in this case, if something happens, probably not important to recover?

Reply:
let me know your thoughts